### DISCUSSION TITLE
Delay in page transition on mobile

### DISCUSSION THREAD

**Discussion by kooalab** [swup/swup#857] [Help] [answered]
### Describe your problem 🧐

HI there,
I have an issue with a basic page transition, specifically when accessed on mobile devices and in situations where the site is not cached (e.g., in private browsing mode). The website I'm working on : https://koppk.netlify.app

Configuration:

`I have set up Swup with the following configuration:  

    const swup = new Swup({    
    animateHistoryBrowsing: true,
    linkToSelf: "scroll",
    plugins: [
        new SwupHeadPlugin({
            awaitAssets: true,
            persistAssets: true
        }), 
        new SwupScriptsPlugin(), 
        new SwupPreloadPlugin({ preloadVisibleLinks: true }), 
        new SwupDebugPlugin()
    ],
    });   
   

CSS Used:   

The CSS for page transitions is as follows:  
```
 html.is-changing .transition-overlay {
      transition: all 0.6s ease;
       }
html.is-animating.is-leaving .transition-overlay {
    backdrop-filter: blur(20px);
    -webkit-backdrop-filter: blur(20px);
    background-color: rgba(7, 8, 11, 1);
}
html.is-rendering .transition-overlay {
    backdrop-filter: blur(0px);
    -webkit-backdrop-filter: blur(0px);
    background-color: rgba(7, 8, 11, 0.001);
}
.transition-overlay {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: 5;
    background-color: rgba(7, 8, 11, 0.05);
    pointer-events: none;
    z-index: 5;
    }
```

On mobile devices, for the first-time visit to the site when no pages are cached, the transition animation behaves unexpectedly. Specifically, after navigating from the homepage to another page, the transition overlay performs normally, but once it fades out, the homepage remains visible for an additional half a second to a second before abruptly switching to the new page. This delay disrupts the smoothness of the transition. Notably, this issue does not occur on subsequent visits when the pages are cached, where the transitions work as intended.

https://github.com/swup/swup/assets/149314442/59d4dfaa-7f4d-48d0-be67-803a7f944713

### What is your question? 🙋‍♀️

I would appreciate any insights or guidance on resolving this issue. Could this be a problem with Swup's handling of transitions on uncached pages, or is there an adjustment needed in my configuration or CSS?

What I don't understand is that the " html.is-rendering" class should only be applied when the content has been replaced so I don't it why the old main element is visible after the transition....

Thanks in advance for your help !

### URL to minimal reproduction 🔗

Don't have it yet

### Checked all these? 📚

- [X] I have read and searched the [official docs](https://swup.js.org)
- [X] I have checked [discussions](https://github.com/orgs/swup/discussions) and [existing issues](https://github.com/issues?q=is:issue+org:swup) for answers to my question
- [X] I agree to follow this project's [Code of Conduct](https://github.com/swup/.github/blob/master/CODE_OF_CONDUCT.md)
- [X] I have provided all necessary information to the best of my knowledge

**Comment by hirasso** [comment]
Hi @kooalab ! 

I'd try two things:
1. remove the `.is-leaving` class from your out-animation:
```css
html.is-animating .transition-overlay {
    backdrop-filter: blur(20px);
    -webkit-backdrop-filter: blur(20px);
    background-color: rgba(7, 8, 11, 1);
}
```
`.is-animating` should be enough. If that doesn't have an effect, you could also...

2. Check if your issue persists if you set [`awaitAssets`](https://swup.js.org/plugins/head-plugin/#awaitassets) to `false` in `SwupHeadPlugin`

**Reply by kooalab** [reply]
Hi @hirasso,

Thank you for the suggestions. Unfortunately, after trying both:

- Removing the .is-leaving class from the out-animation.
- Setting awaitAssets to false in SwupHeadPlugin.

I'm still facing the same issue where there is a noticeable delay when the transition ends before the content gets replaced on the first page visit. 


Additionally, setting awaitAssets to false seemed to have an adverse effect, as the CSS for the new page isn't fully loaded when the page comes into view, which further degrades the user experience.

At this point, I'm not sure what other steps I could take to fix this delay issue. If you have any further insights or suggestions, they would be greatly appreciated.

**Reply by hirasso** [reply]
Could you setup a stackblitz playground with your site? It shouldn’t be too hard, as it’s an Astro site. That would give us the chance to play around with it and help you debug.

**Reply by kooalab** [reply]
Yes sure I can do that, the thing is, this glitch only happends in the production build, while in dev mode everything works fine.

I have the feeling it's Astro that optimizes some things in production which causes this bug on first visit of a page.....

**Reply by kooalab** [reply]
I've set up a StackBlitz playground for the site at: https://stackblitz.com/~/github.com/kooalab/kooa-debug. 

I've attempted to strip down the project to the bare essentials while still replicating the glitch. As previously mentioned, this issue does not appear during development but manifests only in the production build, affecting both mobile and desktop environments.

To observe the glitch, you'll need to build the project and access it through an incognito window on Chrome. The delay where the old content lingers post-transition is observable during the first page visit.

Your assistance in debugging this issue is immensely appreciated. 🙏🏼

**Reply by kooalab** [reply]
Additionally, I have pushed the changes to a live production environment for more straightforward testing: https://debug-swup.netlify.app/. The same glitch occurs there, and it can be observed under the same conditions – on the first visit in an incognito window on both mobile and PC platforms.

Looking forward to your feedback and thanks again for your support.

**Reply by kooalab** [reply]
I don't know if this helps, but what I've realized is that it executes the content:replace and then it says that it's waiting for 2 assets to load which may lead to finishing the animation before the page is ready....

https://github.com/swup/swup/assets/149314442/098fbc3c-f846-4c4e-8a57-e1a01db9f00b

**Reply by hirasso** [reply]
Seems to be related to the head plugin and Astro, as you already suspected. I'm afraid I have not used `HeadPlugin` with stylesheets specific to pages yet, so no experience here. @daun, I think you have used this feature in the past?

I was able to reliably reproduce the issue in chromium like this:

1. in the dev tools, disable the cache and throttle the network speed to "Fast 3G"
2. run `npm run preview`
3. Navigate between the pages

Observation: There are multiple stylesheets being loaded for the subsequent pages (e.g. `/creation-de-site-internet`). The delayed content replacement seems to happen simultaneously with the new CSS files being loaded.

**Reply by hirasso** [reply]
> I don't know if this helps, but what I've realized is that it executes the content:replace and then it says that it's waiting for 2 assets to load which may lead to finishing the animation before the page is ready...

This actually sounds like a good pointer towards the problem!

**Reply by daun** [reply]
The order in the console logs is a bit misleading. It logs the hook name at the beginning of the hook, however long it takes. So if the head plugin correctly delays the content replacement, it will look like `content:replace` has already happened even though it's actually still running.

**Comment by daun** [comment]
I have a feeling this has to do with the CSS transition defined on the overlay. Can you try this CSS based on the official [overlay demo](https://swup.js.org/getting-started/demos/#overlay-animation)?

```css
.transition-overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 999;
  background-color: rgba(7, 8, 11, 1);
  opacity: 0;
  transform: translate3d(0, -100%, 0);
  pointer-events: none;
}

html.is-changing .transition-overlay {
  transition: transform 400ms ease-in-out;
  opacity: 1;
}

html.is-animating .transition-overlay {
  transform: translate3d(0, 0, 0);
}

html.is-rendering .transition-overlay {
  transform: translate3d(0, 100%, 0);
}
```

**Reply by hirasso** [reply]
@daun I just tried your CSS, without success (the bug persists) let's take a few minutes next week to debug this together, should noone have had an epiphany until then :)

**Reply by daun** [reply]
Does enabling/disabling the backdrop filter do anything? And yes, let‘s debug this together next week!

**Reply by kooalab** [reply]
Hey @daun,

With or without the backdrop filter its the same behavior.... 

Thanks guys, means a lot 🙌🏼

**Comment by hirasso** [comment]
@kooalab I can confirm that this is a bug with swup itself. I've created a PR with a fix in #858

**Reply by kooalab** [reply]
Wow that was incredibly fast, thank you so much for the quick turnaround!

**Comment by hirasso** [comment, marked as answer]
@kooalab this should be fixed in 4.5.1. Could you check it out and report back? Thank you!
